core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -306,8 +311,15 @@ protected SqlValidatorImpl(
       SqlValidatorCatalogReader catalogReader,
       RelDataTypeFactory typeFactory,
       Config config) {
-    this.opTab = Objects.requireNonNull(opTab);
     this.catalogReader = Objects.requireNonNull(catalogReader);
+    SqlOperatorTable readerTable = (SqlOperatorTable) catalogReader;
```

Comment:
Is this explicit cast needed?

Reply:
We need it because catalogReader is of type SqlValidatorCatalogReader which isn't of type SqlOperatorTable. It is assumed however that the instance of catalog reader is of type CalciteCatalogReader, which implements Prepare.CatalogReader which implements SqlOperatorTable.